htmd/parameterization/dihedral.py

```diff
@@ -330,41 +345,91 @@ def _evaluateConstTerms(self):
+            modified_parameters = copy.deepcopy(parameters)
+            atomtypes = self._dihedral_atomtypes[idihed]
+            for term in modified_parameters.dihedral_types[atomtypes]:
+                term.phi_k = 0
+                assert term.per > 0 # Guard from messing up with improp
+                    
+            ff = FFEvaluate(self.molecule, modified_parameters)
+            scan_coords = self._coords[idihed]
+            energies = [ff.calculateEnergies(coords[:, :, 0])['total'] for coords in scan_coords]
+            const_energies.append(np.array(energies))
 
+        return const_energies
+
+    def _calculateTargetEnergies(self, const_energies):
+        target_energies = []
+        for ref_energies, const_energies_dih in zip(self._reference_energies, const_energies):
+            target_energies.append(ref_energies - const_energies_dih)
+        shift = np.min(np.concatenate(target_energies))
+        target_energies = [energies - shift for energies in target_energies]
+        return target_energies
+
+    def _fit(self):
         # Save the initial parameters
         vector = self._paramsToVector(self.parameters)
 
         # Evaluate constant terms
-        self._const_energies = self._evaluateConstTerms()
+        const_energies = self._evaluateConstTerms()
 
         # Evaluate target energies for fitting
-        self._target_energies = []
-        for ref_energies, const_energies in zip(self._reference_energies, self._const_energies):
-            energies = ref_energies - const_energies
-            self._target_energies.append(energies)
-
-        shift = np.min(np.concatenate(self._target_energies))
-        self._target_energies = [energies - shift for energies in self._target_energies]
+        target_energies = self._calculateTargetEnergies(const_energies)
 
         # Check self-consistency of computed energies
-        self._objective(vector)
-        test_energies = zip(self._initial_energies, self._const_energies, self._actual_energies)
+        k0, phi0, offset, n = self._unpackVector(vector)
+        actual_energies = _evaluateScanEnergies(k0, phi0, offset, n, self._angle_values, self.numDihedrals, np.arange(self.numDihedrals))
+        test_energies = zip(self._initial_energies, const_energies, actual_energies)
         for initial_energies, const_energies, actual_enegies in test_energies:
             assert np.allclose(initial_energies, const_energies + actual_enegies, rtol=0, atol=1e-5) # TODO debug
 
         # Zero the initial parameters, so they are not used to start the parameter fitting
         if self.zeroed_parameters:
             vector[:] = 0
 
-        # Optimize the parameters
-        logger.info('Start parameter optimization')
+        # Optimize each dihedral individually
+        logger.info('Start parameter optimization with {} iterations'.format(self.num_iterations))
         start = time.clock()
-        vector = self._optimizeWithRandomSearch(vector)
+        best_loss = None
```

Comment:
This should be factor out into `self._optimizeWithXXX(vector)`

Reply:
Done